add SWITCHING_NOZZLE, exclude any XYZ-axis manipulations from SWITCHING_EXTRUDER

fixoid opened the pull request:
I have switching by servo hotends, but two steppers - one for each, because i use bowden extruders.

SWITCHING_EXTRUDER feature is almost suitable for me, but it is designed for DONDOLO like extruder with only one stepper that changes directions for each tool.

I made little changes in firmware to disable reverse and its works well for me.

P.S. sorry for wrong request that i made few days ago.

fixoid committed:
change in indirection for SWITCHING_EXTRUDER

fixoid committed:
Update Configuration.h

thinkyhead commented:
So, in this case… one of the nozzles moves up and down using a servo?
And otherwise, nothing else differs from a normal setup?
If so, I suggest defining it as a new general extruder option instead.
Something like, `SERVO_NOZZLE`, to indicate that every set of 2 nozzles has a servo.

thinkyhead commented:
So, in this case… one of the nozzles moves up and down using a servo?
And otherwise, nothing else differs from a _normal_ setup?

If so, I suggest defining it as a new general extruder option instead.
Something like, `SERVO_NOZZLE`, to indicate that every set of 2 nozzles has a servo.

Compare this to `SINGLENOZZLE_SWITCHING_EXTRUDER` (#6260) which switches the steppers but not the hotend. The changes in that PR are various, because for example in the LCD menus, before moving one of the E steppers, it first has to send the `T` command to select it. This PR probably has similar side-effects. It may be better to take this idea and use it to extend #6260.

fixoid commented:
In my construction both nozzles are moved by servo like a swing, but for firmware its the same  as your assumption - one of the nozzles moves up and down.

Other settings dont differ from a normal setup.

http://www.thingiverse.com/thing:2260178 - here is the video of this suff working.

And there i did new extruder defination SERVO_HOTENDS like you suggest (it was in 1.1.0-RC8 version).
I think SERVO_NOZZLE sounds better :)

fixoid committed:
Update Configuration.h

fixoid committed:
Update Marlin_main.cpp

fixoid committed:
Update configuration_store.cpp

fixoid committed:
Update stepper_indirection.h

fixoid committed:
Update stepper_indirection.h

fixoid committed:
Update Configuration.h

fixoid commented:
Then I propose to exclude any manipulations with the XYZ axes from the `SWITCHING_EXTRUDER` feature, because the purpose of this feature is to press the filament to the different sides of the stepper gear.

This will allow more flexible customization of the firmware for different cases:

- 2 steppers + 2 nozzles with servo (define `SWITCHING_NOZZLE`) - **tested: works well**
- 1 stepper with servo + 2 nozzles  (define `SWITCHING_EXTRUDER`) - **tested: works well**
- 1 stepper with servo + 2 nozzles with the same servo (define `SWITCHING_EXTRUDER` + `SWITCHING_NOZZLE`, SWITCHING_EXTRUDER_SERVO_NR = SWITCHING_NOZZLE_SERVO_NR, SWITCHING_EXTRUDER_SERVO_ANGLES = SWITCHING_NOZZLE_SERVO_ANGLES )  - **tested: works well**
- 1 stepper with servo + 2 nozzles with another servo (for ex. bowden) (define `SWITCHING_EXTRUDER` + `SWITCHING_NOZZLE`, SWITCHING_EXTRUDER_SERVO_NR != SWITCHING_NOZZLE_SERVO_NR, SWITCHING_EXTRUDER_SERVO_ANGLES != SWITCHING_NOZZLE_SERVO_ANGLES ) - **tested: works well**

Also this should work in combination
- 1 stepper + 1 nozzle (define `SWITCHING_EXTRUDER ` + `SINGLENOZZLE `)
without massive changes, as suggested in #6260

fixoid committed:
Update SanityCheck.h

fixoid committed:
Update Conditionals_LCD.h

thinkyhead commented:
>  I propose to exclude any manipulations with the XYZ axes from the `SWITCHING_EXTRUDER` feature

That feature is for an extruder with 2 nozzles, side-by-side, so XYZ movement is needed when switching extruders to move the activated nozzle to the print position.

> because the purpose of this feature is to press the filament to the different sides of the stepper gear

That is what the servo movement does. The XYZ movement aligns the nozzle.

thinkyhead commented:
I see. It does make sense to reorganize in this way. I think XYZ movement on tool-change may only need to depend on there being 2 or more nozzles.

fixoid committed:
Update Marlin_main.cpp

thinkyhead commented on Marlin/Conditionals_LCD.h:
```diff
@@ -291,33 +291,33 @@
-    #define TOOL_E_INDEX current_block->active_extruder
-  #endif
+    #define TOOL_E_INDEX current_block->active_extruder    
+  #endif 
```
Trailing white-space will get you disinvited to parties.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -9550,9 +9564,9 @@ void tool_change(const uint8_t tmp_extruder, const float fr_mm_s/*=0.0*/, bool n
+            move_nozzle_servo(active_extruder);
           #endif
-
+                  
```
boo!

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -9541,6 +9549,12 @@ void tool_change(const uint8_t tmp_extruder, const float fr_mm_s/*=0.0*/, bool n
 
           #if ENABLED(SWITCHING_EXTRUDER)
+            stepper.synchronize();
+
```
remove this blank line

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -331,9 +331,9 @@
  * Only one type of extruder allowed
  */
-#if (ENABLED(SWITCHING_EXTRUDER) && (ENABLED(SINGLENOZZLE) || ENABLED(MIXING_EXTRUDER))) \
+#if (ENABLED(SWITCHING_EXTRUDER) &&  ENABLED(MIXING_EXTRUDER)) \
```
remove this extra space

fixoid commented:
So, i made few changes and tested all cases that i listed in previous message - all works well including `SWITCHING_EXTRUDER` + `SINGLENOZZLE` features

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -331,9 +331,9 @@
+#if (ENABLED(SWITCHING_EXTRUDER) &&  ENABLED(MIXING_EXTRUDER)) \
   || (ENABLED(SINGLENOZZLE) && ENABLED(MIXING_EXTRUDER))
-    #error "Please define only one type of extruder: SINGLENOZZLE, SWITCHING_EXTRUDER, or MIXING_EXTRUDER."
+    #error "Please define only one type of extruder: SINGLENOZZLE/SWITCHING_EXTRUDER or MIXING_EXTRUDER."
```
The condition can be simplified.
```cpp
#if ENABLED(MIXING_EXTRUDER) && (ENABLED(SWITCHING_EXTRUDER) || ENABLED(SINGLENOZZLE))
```
But maybe we should break this into 2 simple errors:
```cpp
#if ENABLED(MIXING_EXTRUDER)
  #if ENABLED(SWITCHING_EXTRUDER)
    #error "Please select either MIXING_EXTRUDER or SWITCHING_EXTRUDER, not both."
  #elif ENABLED(SINGLENOZZLE)
    #error "MIXING_EXTRUDER is not compatible with SINGLENOZZLE."
  #endif
#endif
```

thinkyhead commented on Marlin/Conditionals_LCD.h:
```diff
@@ -291,33 +291,33 @@
    *
    */
-  #if ENABLED(SINGLENOZZLE)             // One hotend, multi-extruder
+  #if ENABLED(SINGLENOZZLE) || ENABLED(MIXING_EXTRUDER)         // One hotend, one termistor, no XY offset
```
"thermistor"

thinkyhead commented on Marlin/Conditionals_LCD.h:
```diff
@@ -291,33 +291,33 @@
+      #ifndef HOTEND_OFFSET_Z
+        #define HOTEND_OFFSET_Z { 0 }
+      #endif  
+    #endif 
```
```cpp
    #if ENABLED(SWITCHING_NOZZLE) && !defined(HOTEND_OFFSET_Z)
      #define HOTEND_OFFSET_Z { 0 }
    #endif
```
…and watch that trailing whitespace. Replace "`[ \t]+$`" with "".

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -137,12 +137,20 @@
-  //#define HOTEND_OFFSET_Z {0.0, 0.0}
+#endif
+
+// A dual nozzle x-carriege that uses a servo motor for switching nozzles
```
"carriage"

thinkyhead replied:
"carriage" … "servomotor"

thinkyhead commented:
Overall it looks good. Some cosmetic changes are needed, but seems functionally sound. Let me know when you've had a chance to patch it up. At that time I'll rebase, squash, and merge it.

thinkyhead commented:
This feature should also include all the changes to `ultralcd.cpp` at https://github.com/MarlinFirmware/Marlin/pull/6260/files#diff-34f7cdd618621ae7d8c113cf0b86a2f4R2010

fixoid committed:
Update Conditionals_LCD.h

fixoid committed:
Update Conditionals_LCD.h

fixoid committed:
Update Configuration.h

fixoid committed:
Update Configuration.h

fixoid committed:
Update Marlin_main.cpp

fixoid committed:
Update SanityCheck.h

fixoid committed:
Update Configuration.h

fixoid committed:
Update SanityCheck.h

fixoid commented:
I made all сorrections you suggested in reviewes.

I would like to make clear if i have to make some changes `ultralcd.cpp` ?
Sorry, I'm newbie here.

fixoid committed:
Update Conditionals_post.h

fixoid committed:
Update SanityCheck.h

thinkyhead commented:
Looks good. There are a couple more things I want to patch up, then I'll squash it down to one commit and merge it. It's a relatively simple PR, and just makes sense. I'll take care of it within the next 24 hours.

fixoid committed:
replaced move_extruder_servo & move_extruder_servo

fixoid commented:
Made little bugfix in `Marlin_main.cpp`:
In order to avoid premature switching of the servo and double switching of the same servo for SWITCHING_NOZZLE + SWITCHING_EXTRUDER with one servo for both (like SWITCHING_EXTRUDER feature in stable version (Marlin 1.1.1)) configuration.

1. replaced `move_extruder_servo `and `move_nozzle_servo`
2. added `SWITCHING_EXTRUDER_SERVO_NR == SWITCHING_NOZZLE_SERVO_NR` сhecking

thinkyhead commented:
Good point. Ok, I'm going to patch it up a little and we'll see what we've got.

thinkyhead commented:
See #6762…
